### PR TITLE
Windows command and script runner

### PR DESCRIPTION
**Pull request by Kami**
This pull request adds experimental action runner for executing commands on remote Windows hosts.

The runner utilizes and offloads hard-work to `winexe` (http://sourceforge.net/projects/winexe/). In the background, winexe utilizes WMI.

With this runner you can also execute power shell scripts, but they already need to exit on the server. I believe it should also be possible to replicate `run-remote-script` behavior and copy script files to a remote host utilizing Samba protocol.

The "hardest" part is building winexe (if you want to use latest version from git) and configuring Windows server. Luckily most cloud providers already enable WMI by default on the cloud servers so only thing you need to do is to configure the firewall.

(As you may see in the examples, I've tested on on Windows 2008 and Windows 2012 Rackspace cloud servers).

Note: This is purely experimental, if we ever wanted to merge this, I would need to add better error handling, tests, documentation, etc.

## Examples

```bash
python ./st2client/st2client/shell.py run core.windows host=x.y.z password="xxx" cmd="ipconfig /all"

status: succeeded
result: 
{
    "stdout": "Windows IP Configuration

   Host Name . . . . . . . . . . . . : TEST-WINDOWS-20
   Primary Dns Suffix  . . . . . . . : 
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No

Ethernet adapter private0:

   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Citrix PV Ethernet Adapter #1
   Physical Address. . . . . . . . . : BC-76-4E-10-0D-67
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv4 Address. . . . . . . . . . . : 10.209.96.125(Preferred) 
   Subnet Mask . . . . . . . . . . . : 255.255.224.0
   Default Gateway . . . . . . . . . : 
   DHCPv6 IAID . . . . . . . . . . . : 331118158
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-A0-8A-25-BC-76-4E-10-09-3C
   DNS Servers . . . . . . . . . . . : 173.203.4.9
                                       173.203.4.8
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter public0:

   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Citrix PV Ethernet Adapter #0
   Physical Address. . . . . . . . . : BC-76-4E-10-09-3C
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv6 Address. . . . . . . . . . . : 2001:4801:7824:101:be76:4eff:fe10:93c(Preferred) 
   IPv4 Address. . . . . . . . . . . : 23.253.20.140(Preferred) 
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : fe80::def%12
                                       23.253.20.1
   DHCPv6 IAID . . . . . . . . . . . : 264009294
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-A0-8A-25-BC-76-4E-10-09-3C
   DNS Servers . . . . . . . . . . . : 173.203.4.9
                                       173.203.4.8
   NetBIOS over Tcpip. . . . . . . . : Enabled

Tunnel adapter Local Area Connection* 11:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Microsoft Teredo Tunneling Adapter
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Tunnel adapter isatap.{DA3F6EFD-C0F4-4338-B270-8BDB9E9FF0D9}:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Microsoft ISATAP Adapter #2
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes", 
    "stderr": "", 
    "exit_code": 0,
    "result": "Windows IP Configuration

   Host Name . . . . . . . . . . . . : TEST-WINDOWS-20
   Primary Dns Suffix  . . . . . . . : 
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No

Ethernet adapter private0:

   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Citrix PV Ethernet Adapter #1
   Physical Address. . . . . . . . . : BC-76-4E-10-0D-67
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv4 Address. . . . . . . . . . . : 10.209.96.125(Preferred) 
   Subnet Mask . . . . . . . . . . . : 255.255.224.0
   Default Gateway . . . . . . . . . : 
   DHCPv6 IAID . . . . . . . . . . . : 331118158
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-A0-8A-25-BC-76-4E-10-09-3C
   DNS Servers . . . . . . . . . . . : 173.203.4.9
                                       173.203.4.8
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter public0:

   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Citrix PV Ethernet Adapter #0
   Physical Address. . . . . . . . . : BC-76-4E-10-09-3C
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv6 Address. . . . . . . . . . . : 2001:4801:7824:101:be76:4eff:fe10:93c(Preferred) 
   IPv4 Address. . . . . . . . . . . : 23.253.20.140(Preferred) 
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : fe80::def%12
                                       23.253.20.1
   DHCPv6 IAID . . . . . . . . . . . : 264009294
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-A0-8A-25-BC-76-4E-10-09-3C
   DNS Servers . . . . . . . . . . . : 173.203.4.9
                                       173.203.4.8
   NetBIOS over Tcpip. . . . . . . . : Enabled

Tunnel adapter Local Area Connection* 11:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Microsoft Teredo Tunneling Adapter
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Tunnel adapter isatap.{DA3F6EFD-C0F4-4338-B270-8BDB9E9FF0D9}:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : 
   Description . . . . . . . . . . . : Microsoft ISATAP Adapter #2
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes"
}
```

```bash
python ./st2client/st2client/shell.py run core.windows host=x.y.z password="xxx" cmd="powershell.exe Get-Process"

status: succeeded
result: 
{
    "stdout": "Handles  NPM(K)    PM(K)      WS(K) VM(M)   CPU(s)     Id ProcessName          
-------  ------    -----      ----- -----   ------     -- -----------          
     34       5     1016       2772    22     0.02   2732 conhost              
    548      12     2208       4324    44     3.48    388 csrss                
    114       9     1796       4096    38     0.12    440 csrss                
    125       8     1944       4880    45     0.31   2212 csrss                
    100       8     2176       6348    34     0.02   1404 dllhost              
    191      15     4608      11380    57     0.69   1752 dllhost              
     70       7     1516       4624    53     0.02   2248 dwm                  
    537      39    15936      34116   214     0.86   2528 explorer             
      0       0        0         24     0               0 Idle                 
    151      23     7368      14776    84     0.14   2808 LogonUI              
    835      27     4940      12344    42     1.03    540 lsass                
    236      10     3132       6292    30     0.05    548 lsm                  
    308      30    47672      42136   604     0.97    176 mmc                  
    149      17     3312       7916    60     0.05   1928 msdtc                
    612      40    64752      10516   624     1.33   2532 Oobe                 
    262      21    68384      63180   576     0.44   1320 powershell           
    930      52    63320      61552   560     1.25   2284 Rackspace.Cloud.Se...
    168      27    25284      20640   493     0.28   2352 Rackspace.Cloud.Se...
     85       7     1432       4156    54     0.05   2424 rdpclip              
    217      12     4152       7912    33     1.01    524 services             
     33       1      444       1100     4     0.02    304 smss                 
    286      20     7772      13644    92     0.51   1312 spoolsv              
    171       9     5504      11668    44     2.96    616 sppsvc               
    354      14     3648       9220    41     0.58    652 svchost              
    268      16     5120       9124    34     0.87    724 svchost              
    256      14     2936       8532    35     0.31    816 svchost              
    325      15     8784      11524    44     0.76    840 svchost              
    409      17     5264      12560    68     0.37    880 svchost              
   1811    1348   619684     383920  1089   391.59    920 svchost              
    494      18     6036      11028    40     1.78    948 svchost              
    534      38    15152      18188   355     0.81    976 svchost              
    308      32     9220      11848    50     0.30   1216 svchost              
     47       4      932       2720    13     0.00   1356 svchost              
     69       7     1564       4600    29     0.06   1828 svchost              
    104      13     2196       5868    34     0.02   2468 svchost              
    623       0      112        304     3    23.70      4 System               
    118      10     2576       5796    53     0.02   2220 taskhost             
     45       4      936       2732    18     0.00   2572 winexesvc            
     76       9     1444       4404    24     0.05    420 wininit              
     98       9     2076       5532    56     0.27    480 winlogon             
     96       7     1624       4952    30     0.03   2980 winlogon             
    217      15     7016      13252    51     0.66   1096 WmiPrvSE             
    147      10     3716       8120    33     0.89   1904 WmiPrvSE             
    160      28    19500      21032   508     0.67   1860 XenDpriv             
    463      39    40076      49492   565     1.59   1448 XenGuestAgent ", 
    "stderr": "", 
    "exit_code": 0,
    "result": "Handles  NPM(K)    PM(K)      WS(K) VM(M)   CPU(s)     Id ProcessName          
-------  ------    -----      ----- -----   ------     -- -----------          
     34       5     1016       2772    22     0.02   2732 conhost              
    548      12     2208       4324    44     3.48    388 csrss                
    114       9     1796       4096    38     0.12    440 csrss                
    125       8     1944       4880    45     0.31   2212 csrss                
    100       8     2176       6348    34     0.02   1404 dllhost              
    191      15     4608      11380    57     0.69   1752 dllhost              
     70       7     1516       4624    53     0.02   2248 dwm                  
    537      39    15936      34116   214     0.86   2528 explorer             
      0       0        0         24     0               0 Idle                 
    151      23     7368      14776    84     0.14   2808 LogonUI              
    835      27     4940      12344    42     1.03    540 lsass                
    236      10     3132       6292    30     0.05    548 lsm                  
    308      30    47672      42136   604     0.97    176 mmc                  
    149      17     3312       7916    60     0.05   1928 msdtc                
    612      40    64752      10516   624     1.33   2532 Oobe                 
    262      21    68384      63180   576     0.44   1320 powershell           
    930      52    63320      61552   560     1.25   2284 Rackspace.Cloud.Se...
    168      27    25284      20640   493     0.28   2352 Rackspace.Cloud.Se...
     85       7     1432       4156    54     0.05   2424 rdpclip              
    217      12     4152       7912    33     1.01    524 services             
     33       1      444       1100     4     0.02    304 smss                 
    286      20     7772      13644    92     0.51   1312 spoolsv              
    171       9     5504      11668    44     2.96    616 sppsvc               
    354      14     3648       9220    41     0.58    652 svchost              
    268      16     5120       9124    34     0.87    724 svchost              
    256      14     2936       8532    35     0.31    816 svchost              
    325      15     8784      11524    44     0.76    840 svchost              
    409      17     5264      12560    68     0.37    880 svchost              
   1811    1348   619684     383920  1089   391.59    920 svchost              
    494      18     6036      11028    40     1.78    948 svchost              
    534      38    15152      18188   355     0.81    976 svchost              
    308      32     9220      11848    50     0.30   1216 svchost              
     47       4      932       2720    13     0.00   1356 svchost              
     69       7     1564       4600    29     0.06   1828 svchost              
    104      13     2196       5868    34     0.02   2468 svchost              
    623       0      112        304     3    23.70      4 System               
    118      10     2576       5796    53     0.02   2220 taskhost             
     45       4      936       2732    18     0.00   2572 winexesvc            
     76       9     1444       4404    24     0.05    420 wininit              
     98       9     2076       5532    56     0.27    480 winlogon             
     96       7     1624       4952    30     0.03   2980 winlogon             
    217      15     7016      13252    51     0.66   1096 WmiPrvSE             
    147      10     3716       8120    33     0.89   1904 WmiPrvSE             
    160      28    19500      21032   508     0.67   1860 XenDpriv             
    463      39    40076      49492   565     1.59   1448 XenGuestAgent "
}
```

```bash
python ./st2client/st2client/shell.py run core.windows host=x.y.z password="xxx" cmd="powershell.exe Geet-Process"

status: failed
result: 
{
    "stdout": "", 
    "stderr": "Geet-Process : The term 'Geet-Process' is not recognized as the name of a 
cmdlet, function, script file, or operable program. Check the spelling of the 
name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ Geet-Process
+ ~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Geet-Process:String) [], Comman 
   dNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException", 
    "exit_code": 1,
    "result": ""
}
```

## TODO / Open questions

Here are the things which need to be done if we decide we want to finish this and merge experimental version into master.

- [x] Refactor and improve the code
- [x] Add more code for handling error conditions and edge cases and propagate friendly exceptions to the user (invalid credentials, timeout, etc.)
- [ ] Audit code for possible shell injections
- [x] Add tests
- [x] Add documentation
  - [x] Document Windows runners
  - [x] Document prerequisites for Window runners
- [x] Decide on the final runner names. Currently it's `run-windows-cmd` and `run-windows-script`. It's not fully consistent, but it's more explicit and better imo. Maybe that's also the time we should consider changing the names of the existing runners and making the names more explicit and consistent (for backward compatibility reasons we would still need to support old names
- [x] Decide on the best place for "green" utils (non-blocking event friendly version of utility functions). `st2common.util.green_shell` vs `st2common.util.green.shell`, etc.
- [x] Make Windows runners opt-in. Add `--experimental` flag to the register-content script - those new runners are only registered when user opt-in and this flag is provided.
- [ ] Build and host `winexe` packages for supported distros (Ubuntu, Fedora). There are some pre-built packages out there, but we don't just want to use a package from a random website and we want to use latest version of winexe from git.
- [ ] Blog post

### CODE REVIEW

**Comment by Kami** [comment]
I also managed to get a runner which allows users to execute PowerShell scripts on remote Windows hosts to work.

It utilizes samba client (`smbclient`) to transfer script file to the remote host and `winexe` to execute the script.

It mostly works out of the box on Windows Rackspace cloud servers, I only needed to:

1. Configure firewall, allow remote access
2. Allow users to run scripts (`powershell.exe Set-ExecutionPolicy RemoteSigned`)

**Comment by dzimine** [comment]
nice!

curious to know about winexe path vs wmi path?

**Comment by Kami** [comment]
You can think of winexe as a subset of wmi-client. It allows you to execute arbitrary commands on a Windows host whereas wmi-client only allows you to query a remote system for different information over WMI using WQL.

winexe works by using samba protocol and rpc to install and start a remote service on Windows host over which communication is then performed. Allowing you to execute arbitrary commands means you can also query WMI locally on the Windows host.

**Comment by Kami** [comment]
I've added "TODO / Open questions" section to the pull request description which lists things that need to be done if we want to make this a reality.

I think it's reasonable to finish it and do a "staged" roll out. Something along the lines of:

1. First we test it ourselves. Spin a couple of Windows cloud servers on Rackspace and Amazon and hammer it for a while.
2. Once we are confident it's good and stable enough, invite a pilot / test customer or two. We need to make it clear this is a preview and there might / will be issues and rough spots.
3. Release it to the "public", drop "experimental" flag.

Eventually, when we support pinning runners to a particular host, we could create a "native" Windows runner which runs on a Windows host and utilizes psexec (https://www.google.si/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=psexec) or similar.

**Comment by Kami** [comment]
> Decide on the best place for "green" utils (non-blocking event friendly version of utility functions). st2common.util.green_shell vs st2common.util.green.shell, etc.

@lakshmi-kannan @manasdk What's your guy's take on this one/

**Comment by dzimine** [comment]
We should move ahead, by @Kami plan but may be even more agressively.
I only suggest that we put windows actions to `win` or `experimental` package to highlight that this is 'experimental' functionlaity, and merge later (or `win` may even be a way to go).

**Comment by Kami** [comment]
> Decide on the best place for "green" utils (non-blocking event friendly version of utility functions). st2common.util.green_shell vs st2common.util.green.shell, etc.

@lakshmi-kannan @manasdk bump - What's your guy's take on this one?

**Comment by manasdk** [comment]
@Kami - That is an ok location for now. Atmost you could  green a `st2common.util.green` package and put under that but it might be unnecessary.

**Comment by lakshmi-kannan** [comment]
I am fine with st2common.util.green.

**Comment by Kami** [comment]
Alright, I've moved it to `st2common.util.green` package.

**Comment by Kami** [comment]
> Decide on the final runner names. Currently it's run-windows-cmd and run-windows-script. It's not fully consistent, but it's more explicit and better imo. Maybe that's also the time we should consider changing the names of the existing runners and making the names more explicit and consistent (for backward compatibility reasons we would still need to support old names

What about this one? I'm kinda OCD about consistency and things like that, but I can also skip it for now so we can get it merged sooner than latter.

**Comment by lakshmi-kannan** [comment]
>Maybe that's also the time we should consider changing the names of the existing runners and making the names more explicit and consistent (for backward compatibility reasons we would still need to support old names

Do you have a proposal? We can come up with aliases and deprecate old names but needs platform changes. I suck at naming things so please halp out with a proposal :P

**Comment by manasdk** [comment]
we would have to register the same runner with 2 names to avoid deprecation issues. So renaming possible but somewhat painful.

**Comment by Kami** [comment]
I was thinking something along those lines:

* run-windows-cmd
* run-windows-script
* run-local-cmd (or run-local-shell-cmd)
* run-local-script (or run-local-shell-script)
* run-remote-cmd (or run-remote-shell-cmd)
* run-remote-script (or run-remote-shell-script)
* run-python-script
* run-http-request

Or we could also drop the run prefix all together.

**Comment by lakshmi-kannan** [comment]
I am not a fan of 'run' prefix TBH. The field is runner-type :P. I somehow want to get the word 'module' in there to suggest it's a piece of code but it would be verbose. So 'remote-script-module', 'local-script-module',...

Dropping the run prefix makes those sound concise and good. I can be +1 to that.

**Comment by Kami** [comment]
@lakshmi-kannan Yeah, agreed.

**Comment by Kami** [comment]
Alright, I pushed the runner rename changes.

I decided to go with "register runner with multiple names" approach. Eventually, we should remove old and deprecated runner names.

**Comment by Kami** [comment]
PR is now ready to be reviewed.

It's not 100% done yet and it needs some more polishing and final touches, but it works and includes a bunch of refactoring so I would like to get those changes merged into master by the end of the week.

File: docs/source/install/windows_runners.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+below in the command prompt:
+
+.. sourcecode:: bash
+
+    netsh firewall set service RemoteAdmin enable
+
+Configuring the administrator user account
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+|st2| requires an administrator account to be able to upload and run the
```
Comment (dzimine) [inline docs/source/install/windows_runners.rst]: Make it more clear that it's a windows account.

**Comment by dzimine** [comment]
This is very nice, we should make big deal of this.
The only concern is how solid is winexe.

**Comment by Kami** [comment]
@dzimine Yeah, that's why we should release it as experimental / beta (it needs more testing in real-life Windows environments, ideally on different Windows server versions).

**Pull request merged**